[FSDP] Change _create_chunk_dtensor in fsdp/_shard_utils.py to use public API from DTensor

wz337 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #110925
* #110846
* __->__ #110831

This PR:
1) updates _create_chunk_dtensor() in _shard_utils.py to use public APIs from DTensor. This will avoid the global_size calculation error from using DTensor.from_local() for uneven-sharded parameters, as described in https://github.com/pytorch/pytorch/issues/110762 
2) updates test/distributed/fsdp/test_fsdp_dtensor_state_dict.py to include unit test for a model with uneven sharding.

cc. @wanchaol, @fegin 

cc @zhaojuanmao @mrshenli @rohan-varma @awgu @fegin @penguinwu

wz337 committed:
change _create_chunk_dtensor to use public API from DTensor

[ghstack-poisoned]

wz337 committed:
Update on "change _create_chunk_dtensor to use public API from DTensor"




[ghstack-poisoned]

fduwjj commented on torch/distributed/fsdp/_shard_utils.py:
```diff
@@ -177,21 +177,17 @@ def _create_chunk_dtensor(
     # HSDP placements: [Replicate(), Shard(0)]
-    placements = [Replicate() for _ in range(device_mesh.ndim)]
-    placements[-1] = shard_placement  # type: ignore[call-overload]
-    return DTensor.from_local(local_tensor, device_mesh, placements)
```
Do we change the logic of local_tensor here? For the old code, it indicates a local shareded Dtensor. But in the new code, we assume it's always local replicate DTensor?

wz337 replied:
> Do we change the logic of local_tensor here? For the old code, it indicates a local shareded Dtensor. But in the new code, we assume it's always local replicate DTensor?

The input tensor for this function here is actually the global tensor we want to shard. We were sharding it in line 180 - 187 in the previous code path(right above) and we were using `shard_placement._split_tensor()`, which is an internal API inside placement_types.py and we don't want to use it. 

Therefore, instead of doing from_local(local_sharded_tensor, device_mesh, Shard(0)) after splitting the tensor, we are using the public DTensor API to do from_local(global_tensor, device_mesh, Replicate()).redistribute(Shard(0)) to create the DTensor.

fduwjj replied:
I see.. If this is the case, why don't we use the api `distribute_tensor`?

https://github.com/pytorch/pytorch/blob/ddb0c265110cfd053009dee34cbc5151bed8cddc/torch/distributed/_tensor/api.py#L424

wanchaol replied:
distribute_tensor does not have assumption about layout current tensor passed in, so for replicate it need to broadcast, sharding it need to scatter, we don't want to call collective here. Also since we already know the tensor is replicated, `from_local` marking it as replicate make more sense.

fduwjj replied:
@wanchaol Well, I mean this looks like a common DTensor operation, would it be better that we can abstract the common logic somewhere in DTensor or we believe this will only be used here for FSDP?

wz337 committed:
Update on "[FSDP] Change _create_chunk_dtensor in fsdp/_shard_utils.py to use public API from DTensor"


This PR:
1) updates _create_chunk_dtensor() in _shard_utils.py to use public APIs from DTensor. This will avoid the global_size calculation error from using DTensor.from_local() for uneven-sharded parameters, as described in https://github.com/pytorch/pytorch/issues/110762 
2) updates test/distributed/fsdp/test_fsdp_dtensor_state_dict.py to include unit test for a model with uneven sharding.

cc. wanchaol, fegin 

cc zhaojuanmao mrshenli rohan-varma awgu fegin penguinwu

[ghstack-poisoned]

wanchaol commented on torch/distributed/fsdp/_shard_utils.py:
```diff
@@ -177,21 +177,17 @@ def _create_chunk_dtensor(
-    # Each chunk is a view of the input tensor. If the original tensor change, the view will also be changed.
     # We need to explicitly call .detach() to return a new tensor detached from the current graph.
-    local_tensor = tensor_list[rank].clone().detach()
+    tensor = tensor.clone().detach()
```
wondering do we still need this? I think we can delete this line as state_dict won't track autograd

fegin replied:
`clone()` is needed but `detach()` may not be necessary.  However, it doesn't harm to keep `detach()` for explicit show grad is not needed..

wanchaol commented on torch/distributed/fsdp/_shard_utils.py:
```diff
@@ -177,21 +177,17 @@ def _create_chunk_dtensor(
-    placements = [Replicate() for _ in range(device_mesh.ndim)]
-    placements[-1] = shard_placement  # type: ignore[call-overload]
-    return DTensor.from_local(local_tensor, device_mesh, placements)
+    replicate_placements = [Replicate() for _ in range(device_mesh.ndim)]
```
do we have a way to detect the current sharding strategy? i.e. if it's fully shard -> [Shard(0)], if it's HSDP, [Replicate(), Shard(0)]

wz337 replied:
> do we have a way to detect the current sharding strategy? i.e. if it's fully shard -> [Shard(0)], if it's HSDP, [Replicate(), Shard(0)]

Agree that this is more explicit and easier to read. We can update all the use case for FSDP and HSDP in later PR, but for 2D FSDP + TP, we will still need to do something like this.

wanchaol commented on torch/distributed/fsdp/_shard_utils.py:
```diff
@@ -177,21 +177,17 @@ def _create_chunk_dtensor(
+    replicate_placements = [Replicate() for _ in range(device_mesh.ndim)]
+    shard_placements = [Replicate() for _ in range(device_mesh.ndim)]
+    shard_placements[-1] = DShard(0)  # type: ignore[call-overload]
+    shard_placements = tuple(shard_placements)
```
I think we don't need this tuple call

wz337 replied:
Removed in the next PR on top of this.

wanchaol commented on test/distributed/fsdp/test_fsdp_dtensor_state_dict.py:
```diff
@@ -126,12 +150,17 @@ def test_dtensor_sharded_tensor_state_dict_identical(self, offload_to_cpu):
+            # if the ShardedTensor is an empty shard,
+            # then the local tensor of DTensor should be local_tensor=tensor([])
+            if len(v2.local_shards()) == 0:
+                self.assertEqual(v1.to_local().numel(), 0)
```
why we are comparing the DTensor path with sharded tensor path? Can be done later but I think we should probably compare with the full state dict path, otherwise the sharded tensor path might become stickier when we get rid of it?

wanchaol reviewed (approved):
lgtm, a few suggestions inlined, thanks for fixing

fegin commented on test/distributed/fsdp/test_fsdp_dtensor_state_dict.py:
```diff
@@ -46,9 +46,29 @@ def get_input(self):
+        self.net1 = nn.Sequential(nn.Linear(5, 10), nn.ReLU())
+        self.net2 = nn.Sequential(nn.Linear(10, 15), nn.ReLU())
+        self.net3 = nn.Linear(15, 30)
+        self.net4 = nn.Sequential(nn.ReLU(), nn.Linear(30, 5))
```
My original test is pretty funny as there are no Relu for net3. I could not think a reason but a typo. We may want to change it in the future to avoid confusion.

wz337 commented:
@pytorchmergebot merge -i

wz337 commented:
@pytorchmergebot merge

malfet commented:
Closing as merged as  https://github.com/pytorch/pytorch/commit/d9eb5a57aade6a3828cbf151b721c060bb8848df by https://github.com/pytorch/pytorch/actions/runs/6474766763
